packages/pds/src/util/repo.ts

```diff
@@ -1,76 +1,179 @@
 import { CID } from 'multiformats/cid'
-import { CidWriteOp, IpldStore } from '@atproto/repo'
+import {
+  DeleteOp,
+  RecordCreateOp,
+  RecordUpdateOp,
+  RecordWriteOp,
+  Repo,
+} from '@atproto/repo'
+import * as auth from '@atproto/auth'
 import { AtUri } from '@atproto/uri'
 import Database from '../db'
 import SqlBlockstore from '../sql-blockstore'
+import { InvalidRequestError } from '@atproto/xrpc-server'
+import { cidForData } from '@atproto/common'
 
-export type MutationContext = {
-  db: Database
-  blockstore: IpldStore
-  did: string
-  time: string
+export const createRepo = async (
+  dbTxn: Database,
+  did: string,
+  authStore: auth.AuthStore,
+  writes: PreparedCreate[],
+  now: string,
+) => {
+  dbTxn.assertTransaction()
+  const blockstore = new SqlBlockstore(dbTxn, did, now)
+  const writeOps = writes.map((write) => write.op)
+  const repo = await Repo.create(blockstore, did, authStore, writeOps)
+  await dbTxn.db
+    .insertInto('repo_root')
+    .values({
+      did: did,
+      root: repo.cid.toString(),
+      indexedAt: now,
+    })
+    .execute()
 }
 
-export const mutationContext = (
-  db: Database,
+export const writeToRepo = async (
+  dbTxn: Database,
   did: string,
-  time?: string,
-): MutationContext => {
-  time = time || new Date().toISOString()
-  return {
-    db,
-    blockstore: new SqlBlockstore(db, did, time),
-    did,
-    time,
+  authStore: auth.AuthStore,
+  writes: PreparedWrites,
+  now: string,
+) => {
+  dbTxn.assertTransaction()
+  const blockstore = new SqlBlockstore(dbTxn, did, now)
+  const currRoot = await dbTxn.getRepoRoot(did, true)
+  if (!currRoot) {
+    throw new InvalidRequestError(
+      `${did} is not a registered repo on this server`,
+    )
   }
+  const writeOps = writes.map((write) => write.op)
+  const repo = await Repo.load(blockstore, currRoot)
+  await repo
+    .stageUpdate(writeOps)
+    .createCommit(authStore, async (prev, curr) => {
+      const success = await dbTxn.updateRepoRoot(did, curr, prev, now)
+      if (!success) {
+        throw new Error('Repo root update failed, could not linearize')
```

Comment:
This shouldn't ever be reached because of the SELECT FOR UPDATE lock. But we perform the check & throw an internal server error just in case